File: scripts/run_tests.py
#!/usr/bin/env python3
"""
Test Runner for Project Arceus
Runs all automated tests and generates a report
"""
import subprocess
import sys
import json
from pathlib import Path
from datetime import datetime


def run_python_tests():
    """Run Python tests with pytest"""
    print("🐍 Running Python tests...")
    
    # Run pytest with coverage
    result = subprocess.run([
        sys.executable, "-m", "pytest",
        "__tests__",
        "-v",
        "--tb=short",
        "--cov=worker",
        "--cov=scripts",
        "--cov-report=term-missing",
        "--cov-report=json"
    ], capture_output=True, text=True)
    
    print(result.stdout)
    if result.stderr:
        print("Errors:", result.stderr)
    
    return result.returncode == 0


def run_javascript_tests():
    """Run JavaScript/TypeScript tests with Jest"""
    print("\n📦 Running JavaScript tests...")
    
    result = subprocess.run([
        "npm", "test", "--", "--coverage", "--json", "--outputFile=test-results.json"
    ], capture_output=True, text=True)
    
    print(result.stdout)
    if result.stderr:
        print("Errors:", result.stderr)
    
    return result.returncode == 0


def run_integration_tests():
    """Run integration tests (requires services running)"""
    print("\n🔗 Running integration tests...")
    
    # Check if services are running
    try:
        import requests
        response = requests.get("http://localhost:3000", timeout=2)
        frontend_running = response.status_code == 200
    except:
        frontend_running = False
    
    if not frontend_running:
        print("⚠️  Frontend not running. Skipping integration tests.")
        print("   Start with: npm run dev")
        return True
    
    # Run API tests
    result = subprocess.run([
        sys.executable, "-m", "pytest",
        "__tests__/api",
        "-v",
        "-m", "not skip"
    ], capture_output=True, text=True)
    
    print(result.stdout)
    return result.returncode == 0


def generate_test_report():
    """Generate a test report"""
    print("\n📊 Generating test report...")
    
    report = {
        "timestamp": datetime.now().isoformat(),
        "results": {}
    }
    
    # Check Python coverage
    coverage_file = Path("coverage.json")
    if coverage_file.exists():
        with open(coverage_file) as f:
            coverage_data = json.load(f)
            report["results"]["python_coverage"] = {
                "percent": coverage_data.get("totals", {}).get("percent_covered", 0),
                "lines": coverage_data.get("totals", {}).get("num_statements", 0),
                "missing": coverage_data.get("totals", {}).get("missing_lines", 0)
            }
    
    # Check Jest results
    jest_results = Path("test-results.json")
    if jest_results.exists():
        with open(jest_results) as f:
            jest_data = json.load(f)
            report["results"]["javascript_tests"] = {
                "passed": jest_data.get("numPassedTests", 0),
                "failed": jest_data.get("numFailedTests", 0),
                "total": jest_data.get("numTotalTests", 0)
            }
    
    # Save report
    report_path = Path("test-report.json")
    with open(report_path, "w") as f:
        json.dump(report, f, indent=2)
    
    print(f"✅ Test report saved to: {report_path}")
    
    # Print summary
    print("\n📋 Test Summary:")
    if "python_coverage" in report["results"]:
        cov = report["results"]["python_coverage"]
        print(f"   Python Coverage: {cov['percent']:.1f}%")
    if "javascript_tests" in report["results"]:
        js = report["results"]["javascript_tests"]
        print(f"   JavaScript Tests: {js['passed']}/{js['total']} passed")


def main():
    """Run all tests"""
    print("🚀 Starting Project Arceus Test Suite\n")
    
    all_passed = True
    
    # Run Python tests
    if not run_python_tests():
        all_passed = False
        print("❌ Python tests failed")
    else:
        print("✅ Python tests passed")
    
    # Run JavaScript tests
    if not run_javascript_tests():
        all_passed = False
        print("❌ JavaScript tests failed")
    else:
        print("✅ JavaScript tests passed")
    
    # Run integration tests
    if not run_integration_tests():
        all_passed = False
        print("❌ Integration tests failed")
    else:
        print("✅ Integration tests passed")
    
    # Generate report
    generate_test_report()
    
    # Exit with appropriate code
    if all_passed:
        print("\n🎉 All tests passed!")
        sys.exit(0)
    else:
        print("\n💥 Some tests failed!")
        sys.exit(1)


if __name__ == "__main__":
    main() 

File: scripts/run-tests.js
#!/usr/bin/env node

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

console.log('🧪 Starting Automated Test Suite...\n');

// Create screenshots directory if it doesn't exist
const screenshotsDir = path.join(__dirname, '../tests/screenshots');
if (!fs.existsSync(screenshotsDir)) {
  fs.mkdirSync(screenshotsDir, { recursive: true });
}

// Test scenarios
const testScenarios = [
  {
    name: 'API Endpoints',
    command: 'npx playwright test tests/api-endpoints.spec.ts --headed',
    description: 'Testing backend API functionality'
  },
  {
    name: 'Scan Review Flow',
    command: 'npx playwright test tests/scan-review-flow.spec.ts --headed',
    description: 'Testing complete user workflow'
  },
  {
    name: 'Visual Regression',
    command: 'npx playwright test --headed --project=chromium',
    description: 'Taking screenshots for visual testing'
  }
];

async function runTests() {
  let passed = 0;
  let failed = 0;

  for (const scenario of testScenarios) {
    console.log(`\n📋 Running: ${scenario.name}`);
    console.log(`📝 Description: ${scenario.description}`);
    
    try {
      execSync(scenario.command, { 
        stdio: 'inherit',
        cwd: path.join(__dirname, '..')
      });
      console.log(`✅ ${scenario.name}: PASSED`);
      passed++;
    } catch (error) {
      console.log(`❌ ${scenario.name}: FAILED`);
      console.log(`Error: ${error.message}`);
      failed++;
    }
  }

  console.log('\n📊 Test Summary:');
  console.log(`✅ Passed: ${passed}`);
  console.log(`❌ Failed: ${failed}`);
  console.log(`📸 Screenshots saved to: tests/screenshots/`);

  // Generate test report
  const report = {
    timestamp: new Date().toISOString(),
    summary: {
      passed,
      failed,
      total: passed + failed
    },
    screenshots: fs.readdirSync(screenshotsDir)
  };

  fs.writeFileSync(
    path.join(__dirname, '../test-report.json'),
    JSON.stringify(report, null, 2)
  );

  console.log('\n📄 Test report saved to: test-report.json');

  if (failed > 0) {
    process.exit(1);
  }
}

runTests().catch(console.error); 